// store/user.js
import { writable } from "svelte/store";
import { exclude } from 'utils/exclude'

/**
 * 
 * @param {import('axios').AxiosInstance} client 
 */
function _createUser(client) {
  const state = {
    token: localStorage.getItem("ukanlearn_access_token") || null,
    username: null,
    email: null,
    drafts: null,
    historic: null,
    role: null,
    timestamps: {
      createdAt: null,
      updatedAt: null
    }
  };

  const { set, update, subscribe } = writable(state);

  return {
    set,
    update,
    subscribe,
    async fetchMe() {
      try {
        const { data } = await client.get('/user/me')
        update(v => ({
          ...v, ...exclude(data, ["_id", "createdAt, updatedAt", "password"]), timestamps: {
            createdAt: data.createdAt, updatedAt: data.updatedAt
          }
        }))
        return null
      } catch (error) {
        return error
      }
    },
    /**
     * 
     * @param {string} email 
     * @param {string} password 
     */
    async login(email, password) {
      try {
        const r = await client.post('/login', {
          email, password
        })
        if (r.data.access_token) {
          update(n => ({ ...n, token: r.data.access_token }))
          localStorage.setItem('ukanlearn_access_token', r.data.access_token)
        }
        return null
      } catch (error) {
        return error
      }
    }
  };
}

export let user = null;

/**
 * 
 * @param {import('axios').AxiosInstance} client 
 */
export function createUser(client) {
  user = _createUser(client)
}
