// store/user.js
import { writable } from "svelte/store";

/**
 * 
 * @param {import('axios').AxiosInstance} client 
 */
function _createUser(client) {
  const state = {
    token: localStorage.getItem("ukanlearn_access_token") || null,
    username: null,
    email: null,
    drafts: null,
    historic: null,
    role: null,
    timestamps: {
      createdAt: null,
      updatedAt: null
    }
  };

  const { set, update, subscribe, get } = writable(state);

  return {
    set,
    update,
    subscribe,
    get,
    /**
     * 
     * @param {string} email 
     * @param {string} password 
     */
    async login(email, password) {
      try {
        const r = await client.post('/login', {
          email,password
        })
        if (r.data.access_token) {
          update(n => ({...n, token: r.data.access_token}))
          localStorage.setItem('ukanlearn_access_token', r.data.access_token)
        }
        return null
      } catch (error) {
        return error
      }
    }
  };
}

export let user = null;

/**
 * 
 * @param {import('axios').AxiosInstance} client 
 */
export function createUser(client) {
  user = _createUser(client)
}
